extras/envoy-xds/pkg/discovery.go

```diff
@@ -0,0 +1,71 @@
+package pkg
+
+import (
+	"encoding/json"
+	"fmt"
+	_ "github.com/Jille/grpc-multi-resolver"
+	"github.com/flipkart-incubator/dkv/pkg/ctl"
+	"github.com/flipkart-incubator/dkv/pkg/serverpb"
+	"io/ioutil"
+	"log"
+)
+
+type DiscoveryClient struct {
+	client     *ctl.DKVClient
+	configPath string
+}
+
+// Create a connection to DKV Service discovery client which provides a view of entire cluster
+// Cluster comprises of all databases and shards (vBuckets) which are registering to this service discovery group
+func InitServiceDiscoveryClient(discoveryServiceAddr string, configPath string) *DiscoveryClient {
+	client, err := ctl.NewInSecureDKVClient(discoveryServiceAddr, "")
+	if (err != nil) {
+	   log.Panicf("Failed to start Discovery Client %v.", err)
+	}
+	return &DiscoveryClient{client: client, configPath: configPath}
+}
+
+func (c *DiscoveryClient) GetEnvoyConfig() (EnvoyDKVConfig, error) {
+
+	config, err := ioutil.ReadFile(c.configPath)
+	if err != nil {
+		return nil, err
+	}
+
+	kvs := make(map[string]interface{})
+	err = json.Unmarshal(config, &kvs)
+	if err != nil {
+		return nil, err
+	}
+
+	regions, err := c.client.GetClusterInfo("", "", "")
+
+	if err != nil {
+		return nil, err
+	}
+
+	all_endpoints := make(map[string][]string)
+
+	for _, region := range regions {
+		var nodeType string
+		if region.Status == serverpb.RegionStatus_LEADER || region.Status == serverpb.RegionStatus_PRIMARY_FOLLOWER ||
+			region.Status == serverpb.RegionStatus_SECONDARY_FOLLOWER {
+			nodeType = "masters"
+		} else {
+			nodeType = "slaves"
```

Comment:
Shouldn't offline nodes be kept out of the config?

Reply:
offline nodes won't be returned (they get filtered from discovery service itself). but still added the check in case a new status is added / offline nodes are returned due to expectation changes from discovery service, etc